ompi/mca/coll/accelerator/coll_accelerator_alltoall.c

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2014-2017 The University of Tennessee and The University
+ *                         of Tennessee Research Foundation.  All rights
+ *                         reserved.
+ * Copyright (c) 2014-2015 NVIDIA Corporation.  All rights reserved.
+ * Copyright (c) 2022      Amazon.com, Inc. or its affiliates.  All Rights reserved.
+ * Copyright (c) 2024      Triad National Security, LLC. All rights reserved.
+ * Copyright (c) 2024      Advanced Micro Devices, Inc. All Rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+#include "coll_accelerator.h"
+
+#include <stdio.h>
+
+#include "ompi/op/op.h"
+#include "opal/datatype/opal_convertor.h"
+
+/*
+ *	Function:	- alltoall for device buffers using temp. CPU buffer
+ *	Accepts:	- same as MPI_Alltoall()
+ *	Returns:	- MPI_SUCCESS or error code
+ */
+int
+mca_coll_accelerator_alltoall(const void *sbuf, size_t scount,
+                      struct ompi_datatype_t *sdtype,
+                      void *rbuf, size_t rcount,
+                      struct ompi_datatype_t *rdtype,
+                      struct ompi_communicator_t *comm,
+                      mca_coll_base_module_t *module)
+{
+    mca_coll_accelerator_module_t *s = (mca_coll_accelerator_module_t*) module;
+    ptrdiff_t sgap, rgap;
+    char *rbuf1 = NULL, *sbuf1 = NULL, *rbuf2 = NULL;
+    int sbuf_dev, rbuf_dev;
+    size_t sbufsize, rbufsize;
+    int rc;
+    int comm_size = ompi_comm_size(comm);
+
+    sbufsize = opal_datatype_span(&sdtype->super, scount, &sgap);
+    rc = mca_coll_accelerator_check_buf((void *)sbuf, &sbuf_dev);
+    if (rc < 0) {
+        return rc;
+    }
+    if ((MPI_IN_PLACE != sbuf) && (rc > 0) &&
+        (sbufsize <= (size_t)mca_coll_accelerator_alltoall_thresh)) {
+        sbuf1 = (char*)malloc(sbufsize * comm_size);
+        if (NULL == sbuf1) {
+            return OMPI_ERR_OUT_OF_RESOURCE;
+        }
+        mca_coll_accelerator_memcpy(sbuf1, MCA_ACCELERATOR_NO_DEVICE_ID, sbuf, sbuf_dev,
```

Comment:
This is not correct, as it copies only the data for rank 0 into the CPU-bound send buffer.